modules/ipc/include/hephaestus/ipc/zenoh/service.h

```diff
@@ -4,30 +4,203 @@
+    std::unique_lock<std::mutex> lock(m);
+    reply_messages.emplace_back(ServiceResponse<ReplyT>{
+        .topic = topic_name, .value = static_cast<std::string>(sample->get_payload().as_string_view()) });
+  } else {
+    CHECK(sample->get_encoding() == Z_ENCODING_PREFIX_EMPTY)
+        << "Encoding for binary types should be Z_ENCODING_PREFIX_EMPTY";
+    auto payload = sample->get_payload();
+    // NOLINTNEXTLINE(cppcoreguidelines-pro-type-reinterpret-cast)
+    std::span<const std::byte> buffer(reinterpret_cast<const std::byte*>(payload.start), payload.len);
+    ReplyT reply_deserialized;
+    serdes::deserialize(buffer, reply_deserialized);
+    std::unique_lock<std::mutex> lock(m);
+    reply_messages.emplace_back(topic_name, std::move(reply_deserialized));
+  }
+}
+}  // namespace internal
+
+template <typename RequestT, typename ReplyT>
+Service<RequestT, ReplyT>::Service(SessionPtr session, TopicConfig topic_config, Callback&& callback)
+  : session_(std::move(session)), topic_config_(std::move(topic_config)), callback_(std::move(callback)) {
+  internal::checkTemplatedTypes<RequestT, ReplyT>();
+
+  auto query = [this](const zenohc::Query& query) mutable {
+    LOG(INFO) << fmt::format("Received query from '{}'", query.get_keyexpr().as_string_view());
+
+    auto reply = this->callback_(internal::deserializeRequest<RequestT>(query));
+    DLOG(INFO) << fmt::format("Replying to '{}'", query.get_keyexpr().as_string_view());
+    zenohc::QueryReplyOptions options;
+    if constexpr (std::is_same_v<ReplyT, std::string>) {
+      options.set_encoding(zenohc::Encoding{ Z_ENCODING_PREFIX_TEXT_PLAIN });  // Update encoding.
+      query.reply(this->topic_config_.name, reply, options);
+    } else {
+      options.set_encoding(zenohc::Encoding{ Z_ENCODING_PREFIX_EMPTY });  // Update encoding.
+      auto buffer = serdes::serialize(reply);
+      DLOG(INFO) << "Replying with buffer of size: " << buffer.size();
+      DLOG(INFO) << fmt::format("Payload: {}", fmt::join(buffer, ","));
+      query.reply(this->topic_config_.name, std::move(buffer), options);
+    }
+  };
+
+  queryable_ = expectAsUniquePtr(
+      session_->zenoh_session.declare_queryable(topic_config_.name, { std::move(query), []() {} }));
+}
+
+template <typename RequestT, typename ReplyT>
+auto callService(Session& session, const TopicConfig& topic_config, const RequestT& request,
+                 const std::optional<std::chrono::milliseconds>& timeout /*= std::nullopt*/)
+    -> std::vector<ServiceResponse<ReplyT>> {
+  internal::checkTemplatedTypes<RequestT, ReplyT>();
+
+  LOG(INFO) << fmt::format("Calling service on '{}'", topic_config.name);
+  std::mutex m;
+  std::condition_variable done_signal;
+  std::vector<ServiceResponse<ReplyT>> reply_messages;
+
+  auto on_reply = [&reply_messages, &m, &topic_config](zenohc::Reply&& reply) {
+    internal::onReply(std::move(reply), reply_messages, m, topic_config.name);
```

Comment:
```suggestion
    internal::onReply(std::forward<zneohc::Reply>(reply), reply_messages, m, topic_config.name);
```

Reply:
in this case you are right it shouldn't matter, but in general you don't know the type of `reply` as it could be an lvalue so for intermediate step you use forward and not worry

Reply:
ah interesting, ok no letr's keep std::move